sale_milestone_profile_invoicing/models/sale_order_line.py

```diff
@@ -0,0 +1,146 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from odoo import api, fields, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    amount_delivered_from_task = fields.Monetary(
+        string='Delivered from task',
+        compute='_compute_amount_delivered_from_task',
+        currency_field='currency_id',
+    )
+    amount_invoiced_from_task = fields.Monetary(
+        string='Invoiced from task',
+        compute='_compute_amount_invoiced_from_task',
+        currency_field='currency_id',
+    )
+    amount_delivered_from_task_company_currency = fields.Monetary(
+        string='Delivered from task (company currency)',
+        compute='_compute_amount_delivered_from_task',
+        currency_field='company_currency_id',
+    )
+    amount_invoiced_from_task_company_currency = fields.Monetary(
+        string='Invoiced from task (company currency)',
+        compute='_compute_amount_invoiced_from_task',
+        currency_field='company_currency_id',
+    )
+    company_currency_id = fields.Many2one(
+        related='company_id.currency_id',
+        store=True
+    )
+
+    is_company_currency = fields.Boolean(
+        compute='_compute_is_company_currency'
+    )
+
+    @api.multi
+    @api.depends('currency_id', 'company_currency_id')
+    def _compute_is_company_currency(self):
+        for line in self:
+            line.is_company_currency = (
+                line.company_currency_id == line.currency_id)
+
+    @api.multi
+    def _get_timesheet_for_amount_calculation(self, only_invoiced=False):
+        """Return all timesheet line related to the sale order line."""
+        self.ensure_one()
+        if not self.task_id:
+            self.amount_delivered_from_task = 0
+            return []
+        tasks_linked_to_line = self.env['project.task'].search(
+            [('sale_line_id', '=', self.id)]
+        )
+        tasks = self.env['project.task'].search(
+            [('id', 'child_of', tasks_linked_to_line.ids)]
+        )
+        ts = tasks.mapped('timesheet_ids').filtered('employee_id')
+        if only_invoiced:
+            ts = ts.filtered('timesheet_invoice_id')
```

Comment:
do we need filter them on state like 'canceled'

Reply:
I think we do, good catch, I updated that filter.